### PR TITLE
make SingleQubitCliffordGate immutable singletons and use it in qubit_characterizations for a 37% speedup

### PR DESCRIPTION
**Pull request by NoureldinYosri**
SingleQubitCliffordGates represents the 24 gates in the single qubit clifford group. Some of the operations this class implements are expensive computation but at the same time are properties of each of the 24 operators.

turning those expensive computations into cached properties is benificial for performance but for that the objects need to be immutable.

one of the places that heavily relies on single qubit cliffords is `qubit_characterizations.py` which implemented the single qubit clifford algebra from scratch by falling on to the matrix representation and doing matrix multiplication and inversion (for computing the adjoint) this led to a bottleneck while creating circuits (e.g. for example _create_parallel_rb_circuit for 50 qubits and a 1000 gates takes $3.886s$). using SingleQubitCliffordGates instead and using `merged_with` operation which maps two cliffords onto the clifford equaivalent to their composition leads to $2.148s$, with most of those 2s spent in `Moment.__init__` which will be the target of the next PR.

I also made the 24 cliffords singleton, since there is no point in creating new object which won't have any of the cached properties.
 
part of https://github.com/quantumlib/Cirq/issues/6389

### CODE REVIEW

File: cirq-core/cirq/ops/clifford_gate.py [hunk 1]
```diff
@@ -356,6 +358,8 @@ def _get_sqrt_map(
 class CliffordGate(raw_types.Gate, CommonCliffordGates):
     """Clifford rotation for N-qubit."""
 
+    _clifford_tableau: qis.CliffordTableau
```
Comment (pavoljuhas) [inline cirq-core/cirq/ops/clifford_gate.py:361]: Nit: We can move this to line 382 as
```
        self._clifford_tableau: qis.CliffordTableau
```

**Review by pavoljuhas — APPROVED** [review]
LGTM with a few minor suggestions.

File: cirq-core/cirq/ops/clifford_gate.py [hunk 2]
```diff
@@ -457,6 +467,7 @@ def _act_on_(
         return NotImplemented
 
 
+@dataclass(frozen=True, init=False, eq=False, repr=False)
```
Comment (maffoo) [inline cirq-core/cirq/ops/clifford_gate.py:470]: What does making this a dataclass actually do here with all these options set to False?

File: cirq-core/cirq/qis/clifford_tableau.py [hunk 3]
```diff
@@ -652,3 +652,6 @@ def measure(
         self, axes: Sequence[int], seed: 'cirq.RANDOM_STATE_OR_SEED_LIKE' = None
     ) -> List[int]:
         return [self._measure(axis, random_state.parse_random_state(seed)) for axis in axes]
+
+    def __hash__(self) -> int:
```
Comment (maffoo) [inline cirq-core/cirq/qis/clifford_tableau.py:656]: consider making this a cached_method as well.

File: cirq-core/cirq/experiments/qubit_characterizations_test.py [hunk 4]
```diff
@@ -73,12 +73,12 @@ def check_distinct(unitaries):
...
             assert is_pauli(u @ p @ u.conj().T), str(u)
 
-    # Check that XZ decomposition has at most one X gate per clifford.
+    # Check that XZ decomposition has at most one X gate and one Z gate per clifford.
     for gates in cliffords.c1_in_xz:
-        num_x = len([gate for gate in gates if isinstance(gate, cirq.XPowGate)])
-        num_z = len([gate for gate in gates if isinstance(gate, cirq.ZPowGate)])
-        assert num_x + num_z == len(gates)
+        num_x = len([gate for gate in gates if gate == cirq.ops.SingleQubitCliffordGate.X])
+        num_z = len([gate for gate in gates if gate == cirq.ops.SingleQubitCliffordGate.Z])
```
[the remark this answers is not in the document]
Comment (NoureldinYosri) [reply cirq-core/cirq/experiments/qubit_characterizations_test.py]: the group has X^{0.5, -0.5, 0} and Z^{1, 0.5, -0.5} so the check has to have sqrtX, and adjointXsqrt. added

File: cirq-core/cirq/ops/clifford_gate.py (hunk 2 above)
Comment (NoureldinYosri) [reply cirq-core/cirq/ops/clifford_gate.py:470]: makes the class frozen while not overriding any method. the default parameters will create an `__init__`, `__eq__` and `__repr__` but we already implement these.

File: cirq-core/cirq/ops/clifford_gate.py [hunk 5]
```diff
@@ -761,6 +777,7 @@ def commutes_with_pauli(self, pauli: Pauli) -> bool:
         to, flip = self.pauli_tuple(pauli)
         return to == pauli and not flip
 
+    @functools.cache
```
[the remark this answers is not in the document]
Comment (NoureldinYosri) [reply cirq-core/cirq/ops/clifford_gate.py]: thanks for the suggestion, changed to cached_method

File: cirq-core/cirq/qis/clifford_tableau.py (hunk 3 above)
Comment (NoureldinYosri) [reply cirq-core/cirq/qis/clifford_tableau.py:657]: done

Comment (NoureldinYosri) [reply cirq-core/cirq/qis/clifford_tableau.py:657]: done

File: cirq-core/cirq/experiments/qubit_characterizations_test.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (pavoljuhas) [reply cirq-core/cirq/experiments/qubit_characterizations_test.py]: I see.  Let us then leave it as it was with a fancier sum-counting of the X and Z gates.

File: cirq-core/cirq/experiments/qubit_characterizations.py [hunk 6]
```diff
@@ -794,10 +787,14 @@ def _single_qubit_gates(
 
 
 def _single_qubit_cliffords() -> Cliffords:
```
[the remark this answers is not in the document]
Comment (NoureldinYosri) [reply cirq-core/cirq/experiments/qubit_characterizations.py:791]: done

**Pull request merged**